symfit/core/fit.py

```diff
@@ -133,23 +170,98 @@ def _init_from_dict(self, model_dict):
 
         :param model_dict: dict of (dependent_var, expression) pairs.
         """
-        sort_func = lambda symbol: str(symbol)
-        self.model_dict = OrderedDict(sorted(model_dict.items(), key=lambda i: sort_func(i[0])))
-        self.dependent_vars = sorted(model_dict.keys(), key=sort_func)
+        sort_func = lambda symbol: symbol.name
+        self.model_dict = OrderedDict(sorted(model_dict.items(),
+                                             key=lambda i: sort_func(i[0])))
+        # Everything at the bottom of the toposort is independent, at the top
+        # dependent, and the rest interdependent.
+        ordered = list(toposort(self.connectivity_mapping))
+        independent = ordered.pop(0)
+        self.dependent_vars = sorted(ordered.pop(-1), key=sort_func)
+        self.interdependent_vars = sorted(
+            [item for items in ordered for item in items],
+            key=sort_func
+        )
+        # `independent` contains both params and vars, needs to be separated
+        self.params = sorted(
+            [s for s in independent if isinstance(s, Parameter)],
```

Comment:
Has been abandoned for now :P. It is not useful when it comes to sorting out the interdependence because it only deals with one `Expr` at a time, and since the Parameters are already present at the lowest level of the toposort it makes more sense to just take them from there.

Reply:
IIRC that function is still in the code and used, but fair enough.